pkg/upgrade/v1alpha2/kubernetes.go

```diff
@@ -0,0 +1,439 @@
+/*
+Copyright 2019 The OpenEBS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"strings"
+	"sync"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	errors "github.com/openebs/maya/pkg/errors/v1alpha1"
+	kclient "github.com/openebs/maya/pkg/kubernetes/client/v1alpha1"
+
+	apis "github.com/openebs/maya/pkg/apis/openebs.io/upgrade/v1alpha1"
+	clientset "github.com/openebs/maya/pkg/client/generated/openebs.io/upgrade/v1alpha1/clientset/internalclientset"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+//TODO: While using these packages UnitTest must be written to corresponding function
+
+// getClientsetFn is a typed function that
+// abstracts fetching of internal clientset
+type getClientsetFn func() (clientset *clientset.Clientset, err error)
+
+// getClientsetFromPathFn is a typed function that
+// abstracts fetching of clientset from kubeConfigPath
+type getClientsetForPathFn func(
+	kubeConfigPath string,
+) (clientset *clientset.Clientset, err error)
+
+// listFn is a typed function that abstracts
+// listing of upgrade task
+type listFn func(
+	cli *clientset.Clientset,
+	namespace string,
+	opts metav1.ListOptions,
+) (*apis.UpgradeTaskList, error)
+
+// getFn is a typed function that
+// abstracts fetching of upgrade task
+type getFn func(
+	cli *clientset.Clientset,
+	namespace, name string,
+	opts metav1.GetOptions,
+) (*apis.UpgradeTask, error)
+
+// createFn is a typed function that abstracts
+// creation of upgrade task
+type createFn func(
+	cli *clientset.Clientset,
+	namespace string,
+	upgradeTask *apis.UpgradeTask,
+) (*apis.UpgradeTask, error)
+
+// deleteFn is a typed function that abstracts
+// deletion of upgradeTasks
+type deleteFn func(
+	cli *clientset.Clientset,
+	namespace, name string,
+	opts *metav1.DeleteOptions,
+) error
+
+// patchFn is a typed function that abstracts
+// to patch upgrade task
+type patchFn func(
+	cli *clientset.Clientset,
+	namespace, name string,
+	pt types.PatchType,
+	data []byte,
+	subresources ...string,
+) (*apis.UpgradeTask, error)
+
+// updateFn is a typed function that abstracts to update
+// upgrade task
+type updateFn func(
+	cli *clientset.Clientset,
+	namespace string,
+	upgradeTask *apis.UpgradeTask,
+) (*apis.UpgradeTask, error)
+
+// make upgrade task clientset as singleton
+var (
+	upgradeTaskClientset *clientset.Clientset
```

Comment:
Let us rename this to `clientsetInstance`

Reply:
Done